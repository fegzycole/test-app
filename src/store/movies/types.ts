export const SET_MOVIES = 'SET_MOVIES';
export const SET_ERROR = 'SET_ERROR';

export interface MovieData {
    imdbID: string;
    Title: string;
    Year: string;
    Type: string;
    Poster: string;
}

export interface MoviesState {
    movies?: MovieData[];
    fetchingMovies: boolean;
    error: string;
}

export interface SetMovies {
    type: typeof SET_MOVIES;
    payload: MovieData[];
}

export interface SetError {
    type: typeof SET_ERROR;
    payload: string;
}

export type MoviesActionTypes = SetMovies | SetError;
