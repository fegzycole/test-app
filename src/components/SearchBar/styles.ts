import Box from '@mui/material/Box';
import IconButton from '@mui/material/IconButton';
import TextField from '@mui/material/TextField';
import { styled } from '@mui/system';
import SearchIcon from '@mui/icons-material/Search';

export const StyledTextField = styled(TextField, {
    name: 'StyledTextField',
    slot: 'Wrapper',
})`
    .MuiOutlinedInput-input {
        padding: 9px 10px;
        border-radius: 0px;
    }

    border-radius: 0px;
    width: 250px;
    color: #28282b;
`;

export const StyledBox = styled(Box, {
    name: 'StyledBox',
    slot: 'Wrapper',
})`
    width: 300px;
    margin: 0 auto;
`;

export const StyledIconButton = styled(IconButton, {
    name: 'StyledButton',
})`
    background-color: #28282b;
    border-radius: 2px;
    margin-left: 10px;
    &:hover {
        background-color: #28282b;
    }
    &.Mui-disabled {
        background: #ccc;
    }
`;

export const StyledSearchIcon = styled(SearchIcon, {
    name: 'StyledSearchIcon',
})`
    color: #fff;
`;
