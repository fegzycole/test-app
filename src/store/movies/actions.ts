import { Dispatch } from 'react';
import { MovieData, SetMovies, SET_MOVIES, SetError, SET_ERROR } from './types';
import axios from 'axios';

const apiKey = process.env.REACT_APP_API_KEY;
const apiUrl = process.env.REACT_APP_API_URL;

export const setMovies = (payload: MovieData[]): SetMovies => {
    return {
        type: SET_MOVIES,
        payload,
    };
};

export const setError = (payload: string): SetError => {
    return {
        type: SET_ERROR,
        payload,
    };
};

export const getMovies = (
    searchText: string,
): ((dispatch: Dispatch<{ type: string; payload?: unknown }>) => Promise<void>) => {
    return async (dispatch) => {
        dispatch(setError(''));

        try {
            const request = await axios.get(`${apiUrl}?s=${searchText}&apikey=${apiKey}`);

            const responseData = request.data.Search;

            const movies: MovieData[] = responseData ? (responseData as MovieData[]) : [];

            dispatch(setMovies(movies));
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
            dispatch(setError('An error occurred, check your internet and try again'));
        }
    };
};
