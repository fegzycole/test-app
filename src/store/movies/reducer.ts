import { MoviesActionTypes, MoviesState, SET_MOVIES, SET_ERROR } from './types';

const initialState: MoviesState = {
    movies: undefined,
    fetchingMovies: false,
    error: '',
};

const moviesReducer = (state: MoviesState = initialState, action: MoviesActionTypes) => {
    switch (action.type) {
        case SET_MOVIES:
            return {
                ...state,
                movies: action.payload,
            };
        case SET_ERROR:
            return {
                ...state,
                error: action.payload,
            };
        default:
            return state;
    }
};

export default moviesReducer;
