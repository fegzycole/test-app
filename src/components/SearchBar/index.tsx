import React from 'react';
import { useDispatch } from 'react-redux';
import { getMovies } from '../../store/movies/actions';
import { StyledBox, StyledTextField, StyledIconButton, StyledSearchIcon } from './styles';

export interface SearchBarProps {
    searchText: string;
    handleChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

export const SearchBar: React.FC<SearchBarProps> = ({ searchText, handleChange }) => {
    const dispatch = useDispatch();

    const handleClick = async () => {
        await dispatch(getMovies(searchText));
    };
    return (
        <StyledBox>
            <StyledTextField value={searchText} onChange={handleChange} placeholder="Enter Text Here" />
            <StyledIconButton onClick={handleClick} disabled={!searchText} data-testid="btn">
                <StyledSearchIcon />
            </StyledIconButton>
        </StyledBox>
    );
};

export default SearchBar;
