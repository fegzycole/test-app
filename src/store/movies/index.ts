import * as Types from './types';
import moviesReducer from './reducer';
import * as Actions from './actions';

export { Types, moviesReducer, Actions };
