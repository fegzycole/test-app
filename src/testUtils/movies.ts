import { MovieData } from '../store/movies/types';

export const mockMovie: MovieData = {
    Title: 'Mock Title',
    Type: 'Mock Type',
    imdbID: 'Mock ID',
    Year: '2022',
    Poster: 'Mock Poster',
};

export const mockMovies: MovieData[] = [mockMovie];
