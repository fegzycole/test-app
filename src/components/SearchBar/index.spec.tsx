import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { useDispatch } from 'react-redux';
import SearchBar, { SearchBarProps } from './index';
import { getMovies } from '../../store/movies/actions';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const mockUseDispatch = useDispatch as jest.Mock<any>;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const mockGetMovies = getMovies as jest.Mock<any>;
const mockDispatch = jest.fn();

const mockProps: SearchBarProps = {
    searchText: '',
    handleChange: jest.fn(),
};

jest.mock('../../store/movies/actions', () => ({
    getMovies: jest.fn(),
}));

describe('<SearchBar />', () => {
    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
    });
    describe('when the search text is empty', () => {
        it('should render correctly', () => {
            const view = render(<SearchBar {...mockProps} />);

            expect(view.container).toMatchSnapshot();
        });
    });

    describe('when the search text is not empty', () => {
        const mockNewProps = { ...mockProps, searchText: 'Mock Search Text' };
        it('should render correctly', () => {
            const view = render(<SearchBar {...mockNewProps} />);

            expect(view.container).toMatchSnapshot();
        });

        describe('when the search button is clicked', () => {
            beforeEach(() => {
                mockUseDispatch.mockImplementation(() => mockDispatch);
                render(<SearchBar {...mockNewProps} />);

                const mockSearchBtn = screen.getByTestId('btn');
                fireEvent.click(mockSearchBtn);
            });

            it('should dispatch an action to fetch the movies matching the search text', () => {
                expect.assertions(2);
                expect(mockDispatch).toHaveBeenCalled();
                expect(mockGetMovies).toHaveBeenCalledWith(mockNewProps.searchText);
            });
        });
    });
});
