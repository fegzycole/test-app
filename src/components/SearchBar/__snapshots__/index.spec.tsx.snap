// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> when the search text is empty should render correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-1pj0buh-StyledBox-wrapper"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-1mj6qlk-MuiFormControl-root-MuiTextField-root-StyledTextField-wrapper"
    >
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          placeholder="Enter Text Here"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-hdw1oc"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-1qhs0lv-MuiButtonBase-root-MuiIconButton-root-StyledButton-root"
      data-testid="btn"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-vr54bh-MuiSvgIcon-root-StyledSearchIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`<SearchBar /> when the search text is not empty should render correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-1pj0buh-StyledBox-wrapper"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-1mj6qlk-MuiFormControl-root-MuiTextField-root-StyledTextField-wrapper"
    >
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r1:"
          placeholder="Enter Text Here"
          type="text"
          value="Mock Search Text"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-hdw1oc"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1qhs0lv-MuiButtonBase-root-MuiIconButton-root-StyledButton-root"
      data-testid="btn"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-vr54bh-MuiSvgIcon-root-StyledSearchIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
